src/5_0/autoImportProviderProject.ts

```diff
@@ -0,0 +1,315 @@
+					compilerOptions,
+					moduleResolutionHost,
+					// @ts-expect-error
+					program.getModuleResolutionCache(),
+					resolveJs,
+				);
+				if (entrypoints) {
+					const real = moduleResolutionHost.realpath?.(packageJson.packageDirectory);
+					const isSymlink = real && real !== packageJson.packageDirectory;
+					if (isSymlink) {
+						symlinkCache.setSymlinkedDirectory(packageJson.packageDirectory, {
+							real,
+							realPath: hostProject.toPath(real),
+						});
+					}
+
+					// @ts-expect-error
+					return mapDefined(entrypoints, (entrypoint) => {
+						const resolvedFileName = isSymlink ? entrypoint.replace(packageJson.packageDirectory, real) : entrypoint;
+						if (!program.getSourceFile(resolvedFileName) && !(isSymlink && program.getSourceFile(entrypoint))) {
+							return resolvedFileName;
+						}
+					});
+				}
+			}
+		},
+
+		create(dependencySelection: PackageJsonAutoImportPreference, hostProject: Project, moduleResolutionHost: ModuleResolutionHost) {
+			if (dependencySelection === PackageJsonAutoImportPreference.Off) {
+				return undefined;
+			}
+
+			const compilerOptions = {
+				...hostProject.getCompilerOptions(),
+				...this.compilerOptionsOverrides,
+			};
+
+			let rootNames: string[] | undefined = this.getRootFileNames(dependencySelection, hostProject, moduleResolutionHost, compilerOptions);
+			if (!rootNames.length) {
+				return undefined;
+			}
+
+			return createAutoImportProviderProject(tsBase, host, createLanguageService, { self: this, hostProject, rootNames, compilerOptions })
+		}	
+	};
+}
+
+function createAutoImportProviderProject(
+	tsBase: typeof import('typescript/lib/tsserverlibrary'),
+	host: LanguageServiceHost,
+	createLanguageService: (host: LanguageServiceHost) => LanguageService,
+	options: AutoImportProviderProjectOptions
+) {
+	const { self, rootNames, compilerOptions, hostProject } = options
+	const ts = tsBase as any;
+	const { some } = ts;
+
+	return {
+		...createProject(
+			tsBase,
+			host,
+			createLanguageService,
+			{
+				projectService: hostProject.projectService,
+				rootNames,
+				currentDirectory: hostProject.currentDirectory,
+				compilerOptions,
+			}
+		),
+
+		rootFileNames: rootNames as undefined | string[],
+
+		hostProject,
+
+		isEmpty() {
+			return !some(this.rootFileNames);
+		},
+	
+		isOrphan() {
+			return true;
+		},
+	
+		updateGraph() {
+			let rootFileNames = this.rootFileNames;
+			if (!rootFileNames) {
+				rootFileNames = self.getRootFileNames(
+					this.hostProject.includePackageJsonAutoImports(),
+					this.hostProject,
+					this.hostProject.getModuleResolutionHostForAutoImportProvider(),
+					this.getCompilationSettings());
+			}
+			this.rootFileNames = rootFileNames;
+			const oldProgram = this.getCurrentProgram();
+			this.program = this.languageService?.getProgram(); 
+			this.dirty = false;
+			if (oldProgram && oldProgram !== this.getCurrentProgram()) {
+				this.hostProject.clearCachedExportInfoMap();
+			}
+		},
```

Comment:
This is the only logic that is not 1:1. 

In summary calling updateGraph should {
- update rootFileNames if not truthy
- set the new program from the language service
- mark project as not dirty (sequential changes in same project won't retrigger updateGraph)
- If the program is not the same then clear the cached export info map.